### PR TITLE
Feature :: concatenate step

### PR DESCRIPTION
**Pull request by vdestraitt**
Introduce a `concatenate` step translated into mongo.
Allows concatenating several columns using a separator of the user's choice.

Closes https://github.com/ToucanToco/vue-query-builder/issues/340

### CODE REVIEW

File: src/components/stepforms/ColumnPicker.vue [hunk 1]
```diff
@@ -41,6 +40,9 @@ export default class ColumnPicker extends Vue {
   @Prop({ default: null })
   dataPath!: string;
 
+  @Prop({ default: true })
+  sync!: boolean;
```
Comment (adimascio) [inline src/components/stepforms/ColumnPicker.vue]: Maybe rename to `syncWithSelectedColumn`? Anyhow, add a docstring to this attribute.

File: src/components/stepforms/ConcatenateStepForm.vue [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+})
+export default class ConcatenateStepForm extends BaseStepForm<ConcatenateStep> {
+  @Prop({
+    type: Object,
+    default: () => ({ name: 'concatenate', columns: [''], separator: '', new_column_name: '' }),
+  })
+  initialStepValue!: ConcatenateStep;
+
+  readonly title: string = 'Concatenate columns';
+  columnPicker = ColumnPicker;
```
Comment (adimascio) [inline src/components/stepforms/ConcatenateStepForm.vue]: Should that be a prop?

File: src/components/stepforms/ConcatenateStepForm.vue [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      // Otherwise, fallback on the default initial value
+      this.editedStep = { ...this.initialStepValue };
+    }
+  }
+
+  get toConcatenate() {
+    if (this.editedStep.columns.length) {
+      return this.editedStep.columns;
+    } else {
+      // return this.defaultColumns;
```
Comment (adimascio) [inline src/components/stepforms/ConcatenateStepForm.vue]: Remove comment

File: src/components/stepforms/ConcatenateStepForm.vue [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      return ['', ''];
```
Comment (adimascio) [inline src/components/stepforms/ConcatenateStepForm.vue]: Hmm, this sounds hackish, why can't we use `['']` or `[]` directly?

File: src/components/stepforms/widgets/ColumnSelector.vue [hunk 5]
```diff
@@ -0,0 +1,69 @@
+<template>
```
Comment (adimascio) [inline src/components/stepforms/widgets/ColumnSelector.vue]: I think this whole file should be removed now that we have the `sync` property on `ColumnPicker`.

File: src/lib/templating.ts [hunk 6]
```diff
@@ -83,6 +83,10 @@ export class PipelineInterpolator implements StepMatcher<S.PipelineStep> {
     return { ...step };
   }
 
+  concatenate(step: Readonly<S.ConcatenateStep>) {
+    return { ...step };
```
Comment (adimascio) [inline src/lib/templating.ts]: Wouldn't it be worth to make the `separator` field templatable?

File: tests/unit/column-picker.spec.ts [hunk 7]
```diff
@@ -70,4 +70,26 @@ describe('Column Picker', () => {
     expect(wrapper.vm.$data.column).toEqual('columnB');
     expect(store.state.vqb.selectedColumns).toEqual(['columnB']);
   });
+
+  it('should not update step when selectedColumn is changed if prop "sync" is false', async () => {
+    const store: Store<any> = setupMockStore({
```
Comment (adimascio) [inline tests/unit/column-picker.spec.ts]: I don't think we need to add the `Store<any>` annotation here.

File: src/components/stepforms/ConcatenateStepForm.vue (hunk 2 above)
Comment (vdestraitt) [reply src/components/stepforms/ConcatenateStepForm.vue]: It conforms to how is built every other component with repeatable widget.
I understand it would change the pattern of how we instantiate components with such a property then?
Maybe it's a change that can e proposed in a specific PR?

File: src/lib/templating.ts (hunk 6 above)
Comment (vdestraitt) [reply src/lib/templating.ts]: No it's ok not to make it templatable. Never saw such a edge case

File: tests/unit/column-picker.spec.ts (hunk 7 above)
Comment (vdestraitt) [reply tests/unit/column-picker.spec.ts]: Need to make typescript happy :)
No quick fix in mind. Maybe we could try to propose something more elegant in another PR ?
